lib/Sema/TypeCheckNameLookup.cpp

```diff
@@ -244,6 +244,90 @@ TypeChecker::lookupUnqualifiedType(DeclContext *dc, DeclName name,
       decls.push_back(cast<TypeDecl>(found.getValueDecl()));
   }
 
+  // If we're looking for a type named `CodingKeys`, then we've got some
+  // additional work to do here. CodingKeys is a special type: although not a
+  // protocol requirement of the Encodable/Decodable protocols, it is
+  // synthesized as part of Encodable/Decodable conformance. Unqualified lookup
+  // may occur before the type is synthesized, though, in which case, we need to
+  // make sure it gets synthesized.
+  //
+  // The unqualified lookups above, however, cannot force Encodable/Decodable
+  // synthesis to occur (since they know nothing about typechecking). Thus, we
+  // also need to find the closest qualified type which might contain a
+  // CodingKeys type and synthesize it.
+  //
+  // If this type is "closer" than the closest unqualified type, then we can
+  // use it.
+  if (name.isSimpleName(Context.Id_CodingKeys)) {
+    SmallVector<TypeDecl *, 1> qualifiedDecls;
+
+    auto *currentDC = dc;
+    while (currentDC) {
+      auto *nominal =
+        currentDC->getAsNominalTypeOrNominalTypeExtensionContext();
+      if (nominal) {
+        // We've found a nominal context to look in. Let's preform a qualified
+        // lookup here (which will cause CodingKeys synthesis if possible).
+        auto lookupOptions = options | NameLookupFlags::ProtocolMembers;
+        auto results = lookupMemberType(nominal,
+                                        nominal->getDeclaredInterfaceType(),
+                                        Context.Id_CodingKeys, lookupOptions);
+        if (results.size() > 0) {
+          for (auto &result : results) {
+            qualifiedDecls.push_back(result.first);
+          }
+
+          break;
+        }
+      }
+
+      currentDC = currentDC->getParent();
+    }
+
+    // If there were any results, we need to merge them with any decls already
+    // found by the unqualified lookup above. Those decls might take precedence
+    // based on their locality.
```

Comment:
Unqualified lookup will always return results at the same "level" today. They're only ordered in the sense that we want deterministic compilation; other than that, there's no reason to merge them in in any way. Either the custom lookup would take precedence over all the other decls (with today's lookup rules) or vice versa.

I feel like there's a better answer that we're missing here, something that normal qualified lookup can hook into without having to go through the TypeChecker's wrapper, but that'd be risky for 4.0 anyway.